### ISSUE TITLE
Strange WISDOM behavior

### ISSUE THREAD

**Issue by OverLordGoldDragon** [FFTW/fftw3#231]
Ported from [pyFFTW Issue](https://github.com/pyFFTW/pyFFTW/issues/306); in summary, `FFTW_WISDOM_ONLY` errors on a call which generates some wisdom but times out, yet without the flag still does "planning" in 2-3 secs as opposed to minutes. Reproduced with:

```python
import pyfftw
import numpy as np

def get_fft_object(x, flags, axis=-1):
    n_fft = x.shape[axis]
    shape = ((n_fft//2 + 1, x.shape[1]) if axis == 0 else
             (len(x), n_fft//2 + 1))

    a = pyfftw.empty_aligned(x.shape, dtype=x.dtype)
    b = pyfftw.empty_aligned(shape, dtype=np.complex64)
    return pyfftw.FFTW(a, b, axes=(axis,), flags=flags, planning_timelimit=120, 
                       threads=8)

x = np.random.randn(400, 200000).astype('float32')

get_fft_object(x, ['FFTW_PATIENT'])()
get_fft_object(x, ['FFTW_PATIENT', 'FFTW_WISDOM_ONLY'])()
```
```python
  File "pyfftw\pyfftw.pyx", line 1280, in pyfftw.pyfftw.FFTW.__cinit__
RuntimeError: No FFTW wisdom is known for this plan.
```

**Comment by hgomersall** [comment]
Just to add to this, I'm fairly sure we're seeing an effect at the `FFTW` level, not in the wrappers. In the case described we're largely just echoing the inner behaviour.

**Comment by OverLordGoldDragon** [comment]
Confirming that without timeout, both `['FFTW_WISDOM_ONLY']` and `['FFTW_PATIENT', 'FFTW_WISDOM_ONLY']` work as intended. Though, planning took 6031 secs for a 10% speedup, which seems rather excessive, then I imagine 'exhaustive' takes a whole day. Planning with 240 secs offered more than x2 speedup, so perhaps there should be something between `PATIENT` and `MEASURE`.

**Comment by matteo-frigo** [comment]
Any chance you can produce a simple C program that shows the behavior?

**Comment by stevengj** [comment]
You are creating the plan using 16- or 32-byte aligned memory via [`pyfftw.empty_aligned`](https://github.com/pyFFTW/pyFFTW/blob/a8fcabde8307b748a7c62dc376d4fb21aeaea6a3/pyfftw/utils.pxi#L173-L214), but then you are *applying* it to potentially unaligned memory allocated with `np.random.randn`.  (By default, numpy arrays are only 8-byte aligned as I understand it.)

The thing to understand is that FFTW's plans are generally specific to the alignment of the data, because that [determines what SIMD instructions we can use](http://fftw.org/doc/SIMD-alignment-and-fftw_005fmalloc.html#SIMD-alignment-and-fftw_005fmalloc).    So, that's why it's failing to create the plan with `WISDOM_ONLY` — the existing wisdom is inapplicable.  However, the wisdom can still accelerate the plan creation, because *parts* of the previous plan may be re-usable even for a differently aligned arrays.

If you want a plan (or wisdom) to be applicable to arrays regardless of alignment, you can use the [FFTW_UNALIGNED flag](http://fftw.org/fftw3_doc/Planner-Flags.html#Planner-Flags) during planning, but in this case the FFT performance may be worse (because certain kinds of SIMD usage will be disallowed).

**Comment by stevengj** [comment]
Hmm, no, you are only planning for the aligned array `a,b`; you are only using `x` for its shape.  (This also means that the resulting plan may be inapplicable to `x`.)

**Comment by stevengj** [comment]
You might want to check the value of pyfftw's [`_simd_alignment`](https://github.com/pyFFTW/pyFFTW/blob/a8fcabde8307b748a7c62dc376d4fb21aeaea6a3/pyfftw/utils.pxi#L45) variable in case it doesn't match FFTW's expectations.

**Comment by stevengj** [comment]
If possible, you could also call the [`fftw_alignment_of` function](http://fftw.org/doc/New_002darray-Execute-Functions.html#index-fftw_005falignment_005fof-1) on your input arrays `a` and `b` to check whether the alignment has changed from FFTW's perspective.

(In short, my first suspicion is that `pyfftw.empty_aligned` may not be doing the right thing.)

**Comment by hgomersall** [comment]
I should prefix this by saying I don't actually care much about this. I'm vaguely interested in what's happening, but it's not exactly pathological.

@stevengj In both cases it's planning for aligned arrays and in both cases with the same alignment. If you try to change the arrays to an unaligned array, it will either error or force a copy (depending on exactly how you do it).

I'd be surprised if the error is with the alignment, the code has been live for perhaps 7 years with no bugs reported. That said, it could easily be something subtle in there. The alignment code is [here](https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/utils.pxi#L173) with the CPU alignment lookup [here](https://github.com/pyFFTW/pyFFTW/blob/master/include/cpu.h). I'd be very interested to know if that's doing something different to what is expected, though I have just noticed that we don't attempt to align for AVX-512, so that could be a potential bug if (a) FFTW does and (b) the OP has a machine that supports AVX-512 and (c) 64-byte alignment is critical for AVX-512.

**Comment by stevengj** [comment]
> In both cases it's planning for aligned arrays and in both cases with the same alignment. 

If they are both 16-byte aligned but FFTW is expecting 32-byte or even 64-byte aligned, then they may not be the same alignment for FFTW. 

It looks like there may be an inherent problem with pfftw's implementation of `empty_aligned`, because it apparently sets the alignment based on the running CPU, but [FFTW sets the minimum alignment based on the architectures for which it is *compiled* to support](https://github.com/FFTW/fftw3/blob/d738232235af6c96bc4a28b07fc867c000696245/kernel/ifftw.h#L133-L145).

**Comment by hgomersall** [comment]
> It looks like there may be an inherent problem with pfftw's implementation of `empty_aligned`, because it apparently sets the alignment based on the detected CPU, but [FFTW sets the minimum alignment based on the architectures for which it is _compiled_ to support](https://github.com/FFTW/fftw3/blob/d738232235af6c96bc4a28b07fc867c000696245/kernel/ifftw.h#L133-L145).

That's interesting. What are the consequences of compiling for, say, AVX-512 and running on a machine that doesn't support it? Given that the runtime alignment lookup should align with what is possible to run on the hardware, it seems that the situation is safe, but I guess the potential problem is if FFTW is compiled for an arch that has a wider alignment and so determines the alignment to be insufficient. Is the consequence a potential slow-down in that it needs to take an unaligned path instead of an aligned path? Or is it as banal as failing a wisdom lookup because the minimum alignments disagree?

**Comment by stevengj** [comment]
> What are the consequences of compiling for, say, AVX-512 and running on a machine that doesn't support it? 

FFTW disables the AVX-512 code at runtime if the CPU doesn't support it.

**Comment by hgomersall** [comment]
> FFTW disables the AVX-512 code at runtime if the CPU doesn't support it.

But does it change the alignment requirement to be 32-bytes?

**Comment by stevengj** [comment]
> But does it change the alignment requirement to be 32-bytes?

`fftw_malloc` still allocates 64-byte aligned memory, but it only *requires* 16-byte alignment for plan applicability as noted above.

**Comment by hgomersall** [comment]
> It only _requires_ the same 16-byte alignment if it was compiled for any SIMD architecture; other alignments may be suboptimal but will still run with the same plans in my recollection. So it only "wastes" alignment in terms of plan applicability if you run on a CPU that doesn't support SIMD at all, which is unlikely these days.

In which case everything should be both safe and architecturally optimal, or am I missing something?

**Comment by stevengj** [comment]
Like I said, please try with the `FFTW_UNALIGNED` flag to see if alignment is the problem.

Otherwise, I'm not sure what is going on; it would be easier to debug if you reproduced the problem with a minimal C program as noted by @matteo-frigo above.

**Comment by hgomersall** [comment]
> Like I said, please try with the `FFTW_UNALIGNED` flag to see if alignment is the problem.
> 
> Otherwise, I'm not sure what is going on; it would be easier to debug if you reproduced the problem with a minimal C program as noted by @matteo-frigo above.

I haven't run anything :) I just pointed @OverLordGoldDragon over here as it looked like an interesting issue. He definitely should explore further with `FFTW_UNALIGNED`.

Forgive my digging on the alignment question. It's important we get this right so if I'm missing something it would great to know!

**Comment by OverLordGoldDragon** [comment]
This issue doesn't concern me any longer as things work out in the end; I'll summarize my findings so far and let maintainers take it from here. In steps,

 1. Run with `FFTW_PATIENT` and `planning_timelimit` on large array; timeout occurs
 2. Run with `FFTW_WISDOM_ONLY`; error
 3. Repeat 1; now it runs in 2-3 secs, as if using wisdom.
 4. I export wisdom, restart Python kernel, then import wisdom, and call with` FFTW_PATIENT` again; again it only takes 2-3 secs
 5. If I _don't_ import wisdom, it takes the full `planning_timelimit`
 6. If `planning_timelimit=None`, it runs for 6000+ secs to completion. _Then_ `FFTW_WISDOM_ONLY` no longer errors.

Basically it works exactly as one'd expect wisdom to. CPU: Intel i7-7700HQ, 4 physical cores / 8 threads.
